### PR TITLE
services: double submit cookie interceptor for channelz

### PR DESCRIPTION
**Pull request by zpencer**
Add this form of XSRF protection to channelz. Non web browser clients
calling channelz must also include the cookie header and metadata
key.

### CODE REVIEW

File: services/src/main/java/io/grpc/services/RequireDoubleSubmitCookieInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+  @Override
+  public <ReqT, RespT> Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call, Metadata headers,
+      ServerCallHandler<ReqT, RespT> next) {
+    String xsrfCookie = null;
+    Iterable<String> cookieHeaders = headers.getAll(COOKIE);
+    if (cookieHeaders == null) {
+      return failCall(call);
+    }
+    for (String cookieHeader: cookieHeaders) {
```
Comment (carl-mastrangelo) [inline services/src/main/java/io/grpc/services/RequireDoubleSubmitCookieInterceptor.java]: nit: space after `:`

File: services/src/main/java/io/grpc/services/RequireDoubleSubmitCookieInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            }
+          }
+        }
+      } catch (IllegalArgumentException e) {
+        log.log(Level.FINE, "Failed to parse cookie header", e);
+        return failCall(call);
+      }
+    }
+    String xsrfHeader = headers.get(headerKey);
+    if (xsrfCookie == null || !Objects.equal(xsrfCookie, xsrfHeader)) {
```
Comment (carl-mastrangelo) [inline services/src/main/java/io/grpc/services/RequireDoubleSubmitCookieInterceptor.java]: `MessageDigest.isEqual` for coolness.

File: services/src/main/java/io/grpc/services/RequireDoubleSubmitCookieInterceptor.java (hunk 1 above)
Comment (zpencer) [reply services/src/main/java/io/grpc/services/RequireDoubleSubmitCookieInterceptor.java]: Done

File: services/src/main/java/io/grpc/services/RequireDoubleSubmitCookieInterceptor.java (hunk 2 above)
Comment (zpencer) [reply services/src/main/java/io/grpc/services/RequireDoubleSubmitCookieInterceptor.java]: Getting the header as `bytes[]` is easy but a `bytes[]` out of a cookie string looks like will require reencoding.

**Pull request merged**